raftstore: gc abnormal snapshots and destroy peer if failed to apply snapshots.

LykxSassinator opened the pull request:
<!--
Thank you for contributing to TiKV!

If you haven't already, please read TiKV's [CONTRIBUTING](https://github.com/tikv/tikv/blob/master/CONTRIBUTING.md) document.

If you're unsure about anything, just ask; somebody should be along to answer within a day or two.

PR Title Format:
1. module [, module2, module3]: what's changed
2. *: what's changed
-->

### What is changed and how it works?
<!--
Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".

For more info, check https://github.com/tikv/tikv/blob/master/CONTRIBUTING.md#linking-issues.
-->

Issue Number: Close #15292

<!--
You could use "commit message" code block to add more description to the final commit message.
For more info, check https://github.com/tikv/tikv/blob/master/CONTRIBUTING.md#format-of-the-commit-message.
-->
What's Changed:

Previously, there were pending tasks to address the scenario where TiKV would panic if applying snapshots failed due to abnormal conditions such as IO errors or unexpected issues.

This pull request resolves the issue by introducing additional traits `tombstone: bool` to `SnapshotApplied`, indicating whether the failure occurred due to abnormal snapshots.
Additionally, the region_id of the abnormal peer will be recorded into the newly added `StoreMeta::damaged_regions`, used to remove the associated peer.
Finally, this peer will be destroyed later through the StoreHeartbeat to PD, which will create a `remove peer` operator to remove the peer safely.


```commit-message
Replace `SnapshotApplied` with `SnapshotApplied { peer_id: u64, tombstone: bool}`. And if `tombstone` == true, the
relative peer will be automatically GCed.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note
<!-- 
Compatibility change, improvement, bugfix, and new feature need a release note.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
-->

```release-note
None.
```

LykxSassinator committed:
raftstore: gc abnormal snapshots and destroy peer if failed to apply a snapshot.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Add logs.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

LykxSassinator committed:
Replace ForceDestroyPeer with Applied {..}

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

glorv commented:
I have 2 questions about this PR:
1. Is it possible to directly retry applying the snapshot if the failure is caused by unexpected(maybe some IO layer) error?
2. Why need to tombstone the peer after applying snapshot failed? Why not just switch the peer status to normal and so the leader should start a new snapshot automatically.

LykxSassinator commented:
> I have 2 questions about this PR:
> 
> 1. Is it possible to directly retry applying the snapshot if the failure is caused by unexpected(maybe some IO layer) error?
> 2. Why need to tombstone the peer after applying snapshot failed? Why not just switch the peer status to normal and so the leader should start a new snapshot automatically.

1. Nope, it's caused by loading abnormal blocks. According to the logs in the issue https://github.com/tikv/tikv/issues/15292 and https://github.com/tikv/tikv/issues/16958, it shows that the given snapshot has abnormal file blocks, causing the failure of the applying.
2. For safety. Tombstone the peer and destroy it will assure that this TiKV node do not have any other remains data / meta data about this peer.

LykxSassinator commented:
By the way, as for the point 2, I agree what u mentioned before. But for safety, this pr takes current implementation.

> Why not just switch the peer status to normal and so the leader should start a new snapshot automatically.

I'll have a try for the `point 2` u mentioned and do some extra tests on it to make comparisons.

overvenus commented:
I don't think we can directly mark it as `Tombstone` or `Normal`, because both options violate the current raft state machine protocol.
* `Tombstone` indicates that the peer has been fully removed, including its Raft membership and data. However, in this case, the peer remains a valid member.
* `Normal` means that the peer has all data up to its last commit index. But, in this case, it does not, as its last commit index has been updated to the snapshot index.

---

There are two ways to fix the panic:

1. Have PD remove this peer via a confchange, which I believe is the simplest solution.
2. Introduce a new RPC to instruct the leader to resend the snapshot, which may change lots of code.

glorv commented:
> Introduce a new RPC to instruct the leader to resend the snapshot, which may change lots of code.

Why need this extra RPC? At the leader side, it will switch the peer's state to normal after finishing send the snapshot. At the follower side, when apply snapshot failed, it is also doable to restore the raft state to its previous state before this snapshot. Thus, If I understand correctly, the leader should start another snapshot without any extra operation?

overvenus commented:
> At the follower side, when apply snapshot failed, it is also doable to restore the raft state to its previous state before this snapshot. 

Do you mean persist the previous state? So it can be restored even after restart TiKV?

That's doable (without a new RPC), but still it adds extra logical to raftstore.

glorv commented:
> Do you mean persisting the previous state so it can be restored even after restarting TiKV?

Yes. But I think just update the peer's state to its previous status should be simpler that tombstone or let pd remove this peer and later add it back.

LykxSassinator commented:
> > Do you mean persisting the previous state so it can be restored even after restarting TiKV?
> 
> Yes. But I think just update the peer's state to its previous status should be simpler than tombstone or let pd remove this peer and later add it back.

Correct me if I'm wrong, but I think that removing the peer and making PD add one fresh peer sounds valid ? As the peer, which panics due to loading the abnormal snapshot, has already advanced its `persisted_index` and `committed_index`, rolling back the persisted state of this peer will violate rules of raft protocol. Right ?

glorv commented:
> Correct me if I'm wrong, but I think that removing the peer and making PD add one fresh peer sounds valid ? As the peer, which panics due to loading the abnormal snapshot, has already advanced its `persisted_index` and `committed_index`, rolling back the persisted state of this peer to a previous one will violate rules of raft protocol. Right ?

You are right, the raft state is reset to the snapshot's state when receives the snapshot msg, so it not easy to revert to the previous state. And makes the raft state revert may cause other side-effects. 

Thus, if it's doable, I prefer the current idea to just destroy the current peer and let newer raft message to trigger create it automatically.

glorv commented:
@LykxSassinator  from https://github.com/tikv/tikv/blob/06eed732641ad47aec44dd58aa3d64e18c1c5ee8/components/raftstore/src/store/fsm/store.rs#L2114-L2130

It seems after tombstoning the peer, the region will drop all following raft messages, thus the peer won't be recovered automatically.

LykxSassinator commented:
> @LykxSassinator from
> 
> https://github.com/tikv/tikv/blob/06eed732641ad47aec44dd58aa3d64e18c1c5ee8/components/raftstore/src/store/fsm/store.rs#L2114-L2130
> 
> It seems after tombstoning the peer, the region will drop all following raft messages, thus the peer won't be recovered automatically.

Yes, thank you for the reminder. It seems that we may need to utilize `ConfChange` for this task.
Anyway, I will investigate further, and this pull request will be put on hold until the necessary changes are implemented and ready for review.

LykxSassinator committed:
Refactor the gc peer with ConfChange request.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

LykxSassinator committed:
Polish codes and supply extra ut.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

Connor1996 commented on components/raftstore/src/store/worker/region.rs:
```diff
@@ -535,18 +539,46 @@ where
+                    .cancel_apply_snapshot(region_id, peer_id);
                 status.swap(JOB_STATUS_FAILED, Ordering::SeqCst);
                 SNAP_COUNTER.apply.fail.inc();
+                // As the snapshot failed, it should be cleared and the related peer should be
```
Why not handle it when `JOB_STATUS_FAILED` in `check_applying_snap`

LykxSassinator commented on components/raftstore/src/store/fsm/peer.rs:
```diff
@@ -1259,9 +1259,18 @@ where
+                    tombstone && self.fsm.peer.peer_id() == peer_id && !self.fsm.peer.is_leader();
+                if apply_snap_failed {
+                    // Send ConfChange to the leader to make the region tombstone the peer.
+                    self.fsm.peer.send_tombstone_peer_msg(self.ctx);
```
Make senses. The absence of messages can result in abnormal peer residue on this node.

Let me dive deeper in it.

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

hbisheng commented:
Out of curiosity and just trying to learn, what are the typical scenarios where applying snapshots may fail? In those cases, would it help if we retry a few times? If we destroy the peer and add a new one, it's possible that the new peer may hit the same issue again, right?

LykxSassinator commented:
> Out of curiosity and just trying to learn, what are the typical scenarios where applying snapshots may fail? In those cases, would it help if we retry a few times? If we destroy the peer and add a new one, it's possible that the new peer may hit the same issue again, right?

Yep. One typical case is that loading snapshots encounters IO errors. In this case, the issue can be attributed to physical errors on the disk or system errors.
And as for whether the retrying is success or not, it depends on the root cause of the IO error. If it's due to physical errors on the disk, retries may not always be successful. But if it's a system error, retries may have a higher chance of success.

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

overvenus commented on components/raftstore/src/store/fsm/peer.rs:
```diff
@@ -3003,11 +3045,7 @@ where
                 self.on_voter_replicated_index_response(msg.get_extra_msg());
             }
             ExtraMessageType::MsgGcPeerRequest => {
```
This message is only used by the 'CheckGcPeer' protocol in raftstorev2. I prefer not to mix them up, as raftstorev2 is not actively being developed and adding this would increase complexity.

LykxSassinator replied:
Agree, it also troubles me a lot. 

Meanwhile, Implementing a retry mechanism seems necessary in case messages are unexpectedly lost

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

LykxSassinator committed:
Polish codes and address comments.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator commented on Cargo.toml:
```diff
@@ -244,6 +244,8 @@ procinfo = { git = "https://github.com/tikv/procinfo-rs", rev = "7693954bd1dd86e
 # [patch.'https://github.com/tikv/rust-rocksdb']
 # rocksdb = { git = "https://github.com/your_github_id/rust-rocksdb", branch = "your_branch" }
+[patch.'https://github.com/pingcap/kvproto']
+kvproto = { git = "https://github.com/LykxSassinator/kvproto", branch = "tombstone_peer" }
```
Will be updated with latest if this pr is approved.

LykxSassinator commented on deny.toml:
```diff
@@ -99,3 +99,7 @@ exceptions = [
 unknown-registry = "deny"
 allow-org = { github = ["tikv", "pingcap", "rust-lang"] }
+# Allow list.
+allow-git = [
```
Will be cleared when this pr is approved.

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

Connor1996 commented on components/raftstore/src/store/peer_storage.rs:
```diff
@@ -795,8 +795,9 @@ where
-                    // TODO: cleanup region and treat it as tombstone.
-                    panic!("{} applying snapshot failed", self.tag,);
+                    // Cleanup region and treat it as tombstone.
+                    warn!("{} applying snapshot failed", self.tag,);
```
```suggestion
                    warn!("{} applying snapshot failed", self.tag);
```

Connor1996 commented on components/raftstore/src/store/fsm/peer.rs:
```diff
@@ -3009,6 +3051,9 @@ where
                     self.on_gc_peer_request(msg);
                 }
             }
+            ExtraMessageType::MsgTombstonePeerRequest => {
```
Tombstone is ambiguous, it actually does remove peer

LykxSassinator replied:
How about `MsgForcelyRemovePeerRequest` ?

Connor1996 replied:
lgtm

LykxSassinator commented on components/raftstore/src/store/fsm/peer.rs:
```diff
@@ -1259,9 +1259,16 @@ where
+            CasualMessage::SnapshotApplied { peer_id, tombstone } => {
                 self.fsm.has_ready = true;
+                // If failed on applying snapshot, it should tomebstone the peer.
+                self.fsm.peer.should_tombstone =
```
To prevent the loss of messages in case of sending failures due to unexpected issues.
The flag `should_tombstone` is used to ensure that the abnormal peer continues to send the TombStone message to the leader until it is removed from the raft group.

Connor1996 replied:
still can send it here

LykxSassinator replied:
FYI, the abnormal peer will stop sending messages once it is destroyed, which is triggered by receiving a `RaftMessage` with the flag `RaftMessage::is_tombstone` set to true.

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

LykxSassinator committed:
Addd extra `unavailable` flag to CheckPeerAvailableResponse.

Mark `unavailable == true` if the abnormal peer failed to apply snapshot and
could not be normarlly removed by ConfChange.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

Connor1996 commented on components/raftstore/src/store/fsm/peer.rs:
```diff
@@ -2255,6 +2262,11 @@ where
+            // If failed on applying snapshot, send ConfChange to the leader to make the
+            // region tombstone the peer.
+            if self.fsm.peer.should_tombstone {
+                self.fsm.peer.send_tombstone_peer_msg(self.ctx);
```
how to avoid sending too many msgs?

Connor1996 commented on components/raftstore/src/store/fsm/peer.rs:
```diff
@@ -2830,11 +2842,24 @@ where
     }
 
-    fn on_availability_response(&mut self, from: &metapb::Peer, msg: &ExtraMessage) {
+    fn on_availability_response(&mut self, msg: RaftMessage) {
```
As it's only used by witness, I don't suggest to change it

LykxSassinator replied:
Yep, that's right. It was reverted.

LykxSassinator replied:
And the abnormal peer will be cleared by the responding message with the `RaftMessage::is_tombstone` == true.

LykxSassinator committed:
Revert "Addd extra `unavailable` flag to CheckPeerAvailableResponse."

This reverts commit cad382d0a94bca8dac8f5541675f29f333b2ccd6.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Polish annotations.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Polish codes and add extra uts.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator commented:
This pr needs extra support from PD side to make the implementation more valid.

Hold until  https://github.com/tikv/pd/pull/8266 is merged.

Connor1996 reviewed (commented):
Please remove this logic for backward compatibility as we can handle abnormal snapshots now
![image](https://github.com/tikv/tikv/assets/13497871/1a88d31a-d303-45cf-b838-ac75c8464487)

LykxSassinator committed:
Merge branch 'master' into clear_if_apply_snap_failed

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Refactor the implementation.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Remove unnecessary deny.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Polish codes.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

overvenus commented on components/raftstore/src/store/fsm/peer.rs:
```diff
@@ -3906,6 +3916,7 @@ where
             );
         })();
         let mut meta = self.ctx.store_meta.lock().unwrap();
+        meta.damaged_regions.remove(&self.fsm.region_id());
```
Please shrink its capacity to free memory.

overvenus commented on tests/failpoints/cases/test_pending_peers.rs:
```diff
@@ -225,3 +225,37 @@ fn test_on_check_busy_on_apply_peers() {
+    // Mock applying snapshot failed on peer 3.
+    fail::cfg("region_apply_snap_io_err", "return").unwrap();
+    pd_client.must_add_peer(region_id, new_peer(3, 3));
+    sleep_ms(1000);
```
Why sleep 1s? Will it be a flakey test?

LykxSassinator replied:
Just one store_heartbeat interval is enough to update meta.

overvenus reviewed (commented):
Rest LGTM

LykxSassinator committed:
Address comments.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

glorv commented on tests/failpoints/cases/test_pending_peers.rs:
```diff
@@ -225,3 +225,37 @@ fn test_on_check_busy_on_apply_peers() {
+    // Mock applying snapshot failed on peer 3.
+    fail::cfg("region_apply_snap_io_err", "return").unwrap();
+    pd_client.must_add_peer(region_id, new_peer(3, 3));
+    sleep_ms(100);
```
Better replace this kind of sleep with `test_util::eventually` to avoid flaky cases.

LykxSassinator replied:
Not appropriate for this, it should wait one region_heartbeat for updating the meta.

glorv reviewed (commented):
rest LGTM

LykxSassinator committed:
Address comments.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

LykxSassinator committed:
Polish codes.

Signed-off-by: lucasliang <nkcs_lykx@hotmail.com>

glorv commented on tests/failpoints/cases/test_pending_peers.rs:
```diff
@@ -247,15 +247,22 @@ fn test_on_apply_snap_failed() {
     // Check that the region is marked as damaged.
-    let stats = cluster.pd_client.get_store_stats(3).unwrap();
+    test_util::eventually(Duration::from_millis(100), Duration::from_secs(1), || {
+        cluster.must_send_store_heartbeat(3);
```
move this line out of this block?

glorv reviewed (approved):
LGTM
